transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,181 @@
+
+/**
+ * A utility class which allows a buffer producer to enqueue arbitrarily sized buffers with promises
+ * and the consumer to dequeue differently sized buffers with the guarantee that when all the bytes in an added
+ * buffer are removed the promise associated with it will be completed when promise passed in during removal of the
+ * last byte completes.
+ *
+ * <p>This functionality is useful for aggregating or partitioning writes into fixed size buffers for framing protocols
+ * such as HTTP2.
+ */
+public final class CoalescingBufferQueue {
+
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(CoalescingBufferQueue.class);
+
+    private final ByteBufAllocator alloc;
+    private final ArrayDeque<Object> promisesAndBufs = new ArrayDeque<Object>();
+    private int readableBytes;
+
+    public CoalescingBufferQueue(ByteBufAllocator alloc) {
+        this.alloc = ObjectUtil.checkNotNull(alloc, "alloc");
+    }
+
+    /**
+     * Add a buffer to the end of the queue.
+     */
+    public void add(ByteBuf buf) {
+        promisesAndBufs.add(ObjectUtil.checkNotNull(buf, "buff"));
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Add a buffer to the end of the queue and associate a promise with it that should be completed when
+     * all it's bytes have been consumed from the queue and written.
+     * @param buf to add to the tail of the queue
+     * @param promise to complete when all the bytes have been consumed and written, can be void.
+     */
+    public void add(ByteBuf buf, ChannelPromise promise) {
+        // buffers are added before promises so that we naturally 'consume' the entire buffer during removal
+        // before we complete it's promise.
+        ObjectUtil.checkNotNull(buf, "buf");
+        ObjectUtil.checkNotNull(promise, "promise");
+        promisesAndBufs.add(buf);
+        if (!promise.isVoid()) {
+            promisesAndBufs.add(promise);
+        }
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Remove a {@link ByteBuf} from the queue with the specified number of bytes. Any added buffer who's bytes are
+     * fully consumed during removal will have it's promise completed when the passed aggregate {@link ChannelPromise}
+     * completes.
+     *
+     * @param bytes the maximum number of readable bytes in the returned {@link ByteBuf}.
+     * @param aggregatePromise used to aggregate the promises for the constituent buffers.
+     * @return a {@link ByteBuf} composed of the enqueued buffers.
+     */
+    public ByteBuf remove(int bytes, ChannelPromise aggregatePromise) {
+        if (bytes < 0) {
+            throw new IllegalArgumentException("bytes cannot be less than 0 : " + bytes);
+        }
+        ObjectUtil.checkNotNull(aggregatePromise, "aggregatePromise");
+
+        // Use isEmpty rather than readableBytes==0 as we may have a promise associated with an empty buffer.
+        if (promisesAndBufs.isEmpty()) {
```

Comment:
Ok that's fine ... but we should document what the behavior is in that case so that it's clear.  Probably worth doing something like this as well (if for nothing else than clarity):

`bytes = min(bytes, readableBytes)`